src/Shared/Model/ArtifactUri.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Model;
+
+using PackageManagers;
+using Polly;
+using Polly.Contrib.WaitAndRetry;
+using Polly.Retry;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+/// <summary>
+/// A record to represent the type, uri, and extension for an artifact associated with a package.
+/// </summary>
+/// <typeparam name="T">The enum to represent the artifact type.</typeparam>
+public record ArtifactUri<T> where T : Enum
+{
+    /// <summary>
+    /// Initializes a new instance of <see cref="ArtifactUri{T}"/>.
+    /// </summary>
+    /// <param name="type">The type of artifact for this <see cref="ArtifactUri{T}"/>.</param>
+    /// <param name="uri">The <see cref="Uri"/> this artifact can be found at.</param>
+    /// <param name="extension">The file extension for the file found at the <paramref name="uri"/>.</param>
+    public ArtifactUri(T type, Uri uri, string? extension = null)
```

Comment:
Why is extension configurable separate from uri? For .tar.gz?

Reply:
This is removed from the most recent commit. Looks like you reviewed an older version.

Reply:
But to answer your question, it wasn't ever being used so it doesn't really matter. But yeah there is the issue with the .tar.gz files still, so if you know of a good solution please let me know!

Reply:
Some untested thoughts:
1. You could take everything after the last '/' and then find the first the first '.' in that and take everything after that '.' as an extension.
2. You could use the Get extension thing below and just have special cases for .tar.* (also possible to have like a .tar.bz2, we shouldn't just restrict to tar.gz.

Reply:
The first one wouldn't work because of file names like this: https://pypi.org/packages/source/r/requests/requests-2.27.1.tar.gz 
So I'll mess around with the second one

Reply:
Had to write my own method to do it. Just pushed.

Reply:
I think that looks like a fine way to do it.